src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Timestamp.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+#if IS_DESKTOP
+using System.Security.Cryptography.Pkcs;
+#endif
+
+namespace NuGet.Packaging.Signing
+{
+    public sealed class Timestamp
+    {
+#if IS_DESKTOP
+
+        /// <summary>
+        /// Upper limit of Timestamp.
+        /// </summary>
+        public DateTimeOffset UpperLimit { get; }
+
+        /// <summary>
+        /// Lower limit of Timestamp.
+        /// </summary>
+        public DateTimeOffset LowerLimit { get; }
+
+        /// <summary>
+        /// Time timestamp was created by the Time Stamp Authority.
+        /// </summary>
+        public DateTimeOffset GeneralizedTime { get; }
+
+        /// <summary>
+        /// Timestamp SignerInfo.
+        /// </summary>
+        public SignerInfo SignerInfo { get; }
+
+
+        /// <summary>
+        /// SignedCms containing a time stamp authority token reponse
+        /// </summary>
+        /// <param name="timestampCms">SignedCms from Time Stamp Authority</param>
+        public Timestamp(SignedCms timestampCms)
+        {
+            SignerInfo = timestampCms?.SignerInfos[0] ?? throw new ArgumentNullException(nameof(timestampCms));
+
+            if (Rfc3161TimestampVerificationUtility.TryReadTSTInfoFromSignedCms(timestampCms, out var tstInfo))
+            {
+                GeneralizedTime = tstInfo.Timestamp;
+
+                var accuracyInMilliseconds = Rfc3161TimestampVerificationUtility.GetAccuracyInMilliseconds(tstInfo);
+                UpperLimit = tstInfo.Timestamp.AddMilliseconds(accuracyInMilliseconds);
+                LowerLimit = tstInfo.Timestamp.AddMilliseconds(-accuracyInMilliseconds);
+            }
+            else
+            {
+                throw new Exception("TODO");
```

Comment:
you could throw `TimestampException(Strings.TimestampFailureInvalidContentType)`

Reply:
Please fix the exception as mentioned by @PatoBeltran